clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/model/CloudMetricProvider.groovy

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright 2016 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.model
+
+interface CloudMetricProvider<T extends CloudMetricDescriptor> {
+
+  /**
+   * Returns the platform of the provider
+   * @return a String, e.g. 'aws', 'gcp'
+   */
+  String getCloudProvider()
+
+  /**
+   * Returns a specific metric descriptor
+   * @param account the account
+   * @param region the region
+   * @param filters a collection of identifiers used to uniquely identify a metric
+   * @return a metric descriptor if one is found; should throw an exception if multiple metric descriptors are found
+   * for the supplied filters
+   */
+  T getMetricDescriptor(String account, String region, Map<String, String> filters)
+
+  /**
+   * Returns a list of metric descriptors matching the supplied filters
+   * @param account the account
+   * @param region the region
+   * @param filters a collection of identifiers used to select a subset of all metrics in the account and region
+   * @return a list of metric descriptors matching the filters
+   */
+  List<T> listMetricDescriptors(String account, String region, Map<String, String> filters)
+
+  /**
+   * Returns a statistic set for the metric descriptor uniquely identified by the supplied filters
+   * @param account the account
+   * @param region the region
+   * @param name the name of the target metric
+   * @param filters a collection of identifiers used to uniquely identify a metric
+   * @param startTime an inclusive timestamp to determine the oldest datapoint to return
+   * @param endTime an exclusive timestamp to determine the newest datapoint to return
+   * @return a CloudMetricStatistics object, describing the statistics with timestamps
+   */
+  CloudMetricStatistics getStatistics(String account, String region, String name, Map<String, String> filters,
```

Comment:
Well, I have to walk back that last comment. 

It makes sense, but I can't find a clean way to implement it. To get the statistics for AWS, I need the following fields to identify the metric:
* account
* region
* metricName
* namespace
* dimensions (an arbitrary map of values)

Then, I need these fields to determine how to generate the data points:
* statistic (average, sum, maximum, etc.)
* period (number of seconds per sample)
* startTime
* endTime

I'm guessing GCM will have a different set of fields, so I am unsure how to go about this. I tried adding a `from` method to the `CloudMetricProvier` interface, i.e.
```java
T from(Map<String, String> parameters)
```
but got myself into trouble implementing the AWS side around knowing which fields should be dimensions and which fields should be for generating data. Everything I tried ended up muddling concerns between the controller, the interfaces, and the provider implementations.

For now, I'm going to leave this as-is. If you or someone else can find a good way to refactor this once another cloud provider gets in the mix, I am all for it.